src/IO/ReadWriteBufferFromHTTP.h

```diff
@@ -401,19 +413,30 @@ namespace detail
                 saved_uri_redirect = uri_redirect;
             }
 
+            if (response.hasContentLength())
+                LOG_DEBUG(log, "Received response with content length: {}", response.getContentLength());
+
             if (withPartialContent() && response.getStatus() != Poco::Net::HTTPResponse::HTTPStatus::HTTP_PARTIAL_CONTENT)
             {
                 /// Having `200 OK` instead of `206 Partial Content` is acceptable in case we retried with range.begin == 0.
                 if (read_range.begin && *read_range.begin != 0)
                 {
                     if (!exception)
+                    {
                         exception = std::make_exception_ptr(Exception(
                             ErrorCodes::HTTP_RANGE_NOT_SATISFIABLE,
-                            "Cannot read with range: [{}, {}]",
+                            "Cannot read with range: [{}, {}] (response status: {}, reason: {})",
                             *read_range.begin,
-                            read_range.end ? *read_range.end : '-'));
+                            read_range.end ? toString(*read_range.end) : "-",
+                            toString(response.getStatus()), response.getReason()));
+                    }
+
+                    /// Retry 200OK
+                    if (response.getStatus() == Poco::Net::HTTPResponse::HTTPStatus::HTTP_OK)
+                        initialization_error = InitializeError::RETRIABLE_ERROR;
```

Comment:
Would simply throwing an exception here work also?
It should be caught in the `nextImpl`. The only thing you need to save in a variable is the error message, and assign it to `exception` if it's not 200, and throw exception with it if it is 200.

Reply:
I though of doing it this way at first, but current way seemed to make it more clear from the code what exactly we want to retry. Also in that try-catch you are referring to, there is no good way to check http status code as `response` variable is not accessible anymore, we can only see http error code `RANGE_IS_NOT_SATISFIABLE` there.

Reply:
As far as I see, you don't need the response anymore if you throw the exception here, it will be caught in the `nextImpl` and everything will be retried correctly.

Reply:
But I agree about it being more clear 👍

Reply:
> As far as I see, you don't need the response anymore if you throw the exception here, it will be caught in the nextImpl and everything will be retried correctly.

Ah, indeed...